docs/prover/approx/hashing.md

```diff
@@ -0,0 +1,182 @@
+
+
+### Example (Imprecision of Modeling)
+
+Whichever distinct values we chose for `x` and `y` in the example below, on the real Keccak function one rule would be violated and one rule would not. In the modeling of the Certora Prover, both rules are violated, since the prover is allowed to "invent" a hash function for each rule and will choose one that violates the property.
+
+```solidity
+// CVL:
+methods { hash(uint) returns (uint) envfree; }
+
+definition x() : uint = 12345678
+definition y() : uint = 87654321
+
+rule hashXLowerOrEqualToHashY {
+	assert hash(x()) <= hash(y());
+}
+
+rule hashXLargerThanHashY {
+	assert hash(x()) > hash(y());
+}
+
+// solidity:
+contract C {
+	function hash(uint x) public returns (bytes32) {
+		return keccak256(x);
+	}
+}
+
+```
+
+
+<!---
+[comment]: # We should also clearly explain the surprising thing the user might see. A clear example of a rule that should pass and doesn't, or a rule that shouldn't pass and does, and an explanation of why.
+--->
+
+## Hashing of Unbounded Data
+
+In the discussion so far we only considered hashes of data whose length is already known before program execution (e.g. a `uint` variable always has 256 bit). Hashing of unbounded data (typically unbounded arrays, like `bytes`, `uint[]`, etc.) requires some extra measures, since their implementation requires loops and the the Certora Prover internally eliminates all loops to other constructs in order to achieve better tractability.
+
+The Certora Prover models unbounded hashing similar to how it eliminates loops. The user specifies an upper length bound up to which unbounded hashing should be modeled precisely (using the CLI option {ref}`--hashing_length_bound`) as well as whether this bound is to be assumed or to be verified (using the CLI option {ref}`--optimistic_hashing`).
+
+We demonstrate how these flags work using the following program snippet.
+
+```solidity
+contract C {
+	mapping(bytes => uint) m;
+	bytes b1, b2, b3;
+	uint u, v, w;
+	// ...
+		require b1.length < 224;
+		m[b1] = u;
+	// ...
+		// no constraints on b2.length
+		m[b2] = v; 
+	// ...
+		m[b3] = v;
+	    assert(b3.length > 300, "b3 is less than 300 bytes long, unexpectedly")
+	// ...
+}
+```
+
+Let us assume that the `--hashing_length_bound` flag is set to 224 (which corresponds to 7 machine words).
+Then, the first hash operation, triggered by the mapping access `m[b1]`, behaves like the hash of a bounded data chunk. The `--optimstic_hashing` flag has no impact on this hash operation.
+Behavior of the second hash operation, triggered by the mapping access `m[b2]`, depends on whether `--optimistic_hashing` is set. 
+If the `--optimistic_hashing` flag is not set, the violation of an internal assertion will be reported by the prover, stating that an chunk of data is being hashed that may exceed the given bound of 224.
```

Comment:
It would be good to include the exact message so that people can hope to find this issue by searching for the error when they see it.

Reply:
alright, done